Dockerfiles for learning locker

lunika opened the pull request:
We want to dockerised the learning locker stack. To do this we have to create the image for learning locker.

The learning locker image follow the tag used by the learning locker team in their github repository. The last version is `v2.6.2` so we will have to create an image with the tag `v2.6.2`

lunika committed:
✨(LRS) create Dockerfile for learning locker

Learning locker has 4 apps but we manage them in one image, it's not
really possible to split them.

lunika committed:
✨(docker-compose) create a docker-compose config for example

To ease the use of our images, we have created a docker-compose
configuration.

jmaupetit commented on README.md:
```diff
@@ -1 +1,71 @@
+
+This repo contains the Dockerfile to build images for [learning locker](https://github.com/LearningLocker/learninglocker), a LRS to store and analyze xAPI statements.
+
+We follow the learning locker's tag, so an image will have the same tag you can found in learning loacker repository.
```
```suggestion
We follow the learning locker's tag, so an image will have the same tag you can found in learning locker repository.
```

jmaupetit commented on README.md:
```diff
@@ -1 +1,71 @@
+
+### Build an nginx image
+
+The nginx image is here as a proxy to communicate with learning locker. It will be useful in your development stack, in a docker-compose configuration for example.
```
```suggestion
The `nginx` image is used as a reverse-proxy to learning locker. It will be useful in your development stack (_e.g._ a docker-compose configuration).
```

jmaupetit commented on README.md:
```diff
@@ -1 +1,71 @@
+
+The nginx image is here as a proxy to communicate with learning locker. It will be useful in your development stack, in a docker-compose configuration for example.
+
+You have to set the in the `VERSION` environment variable the tag version you want to use and then use the `build-nginx` task in our Makefile
```
```suggestion
You have to set the `VERSION` environment variable for the tag version you want to use and then use the `build-nginx` task in our Makefile:
```

jmaupetit replied:
Is this an nginx release or learning locker release?

jmaupetit commented on README.md:
````diff
@@ -1 +1,71 @@
+node worker/dist/server
+```
+
+You can also execute a CLI application, to do that you can execute this command inside your container:
````
```suggestion
You can also use the CLI by executing the following command from your container:
```

jmaupetit commented on learninglocker/Dockerfile:
```diff
@@ -0,0 +1,30 @@
+ARG LL_VERSION
+
+# Download release and extract it in /opt/learninglocker
+RUN cd /tmp && curl --compressed -L -o "learninglocker-$LL_VERSION.tar.gz" "https://github.com/LearningLocker/learninglocker/archive/$LL_VERSION.tar.gz" \
```
```suggestion
RUN cd /tmp && curl --compressed -L -o "learninglocker-${LL_VERSION}.tar.gz" "https://github.com/LearningLocker/learninglocker/archive/${LL_VERSION}.tar.gz" \
```

jmaupetit commented on learninglocker/Dockerfile:
```diff
@@ -0,0 +1,30 @@
+
+# Download release and extract it in /opt/learninglocker
+RUN cd /tmp && curl --compressed -L -o "learninglocker-$LL_VERSION.tar.gz" "https://github.com/LearningLocker/learninglocker/archive/$LL_VERSION.tar.gz" \
+    && tar -xzf "learninglocker-$LL_VERSION.tar.gz" \
```
```suggestion
    && tar -xzf "learninglocker-${LL_VERSION}.tar.gz" \
```

jmaupetit commented on learninglocker/Dockerfile:
```diff
@@ -0,0 +1,30 @@
+RUN cd /tmp && curl --compressed -L -o "learninglocker-$LL_VERSION.tar.gz" "https://github.com/LearningLocker/learninglocker/archive/$LL_VERSION.tar.gz" \
+    && tar -xzf "learninglocker-$LL_VERSION.tar.gz" \
+    && rm learninglocker-*.tar.gz \
+    && mv learninglocker* /opt/learninglocker
```
Use `/app` instead of `/opt`

jmaupetit commented on learninglocker/Dockerfile:
```diff
@@ -0,0 +1,30 @@
+
+# Install learning locker
+RUN cd /opt/learninglocker \
+    && touch .env \
```
Please add a comment for this trick.

jmaupetit commented on nginx/Dockerfile:
```diff
@@ -0,0 +1,10 @@
+
+ENTRYPOINT ["/entrypoint.sh"]
+
+CMD ["nginx", "-g", "daemon off;"]
```
I don't think we need this image at all. As you suggested IRL, we can use the official nginx image combined with `envsubst`.

jmaupetit commented on learninglocker/Dockerfile:
```diff
@@ -0,0 +1,30 @@
+    && rm learninglocker-*.tar.gz \
+    && mv learninglocker* /opt/learninglocker
+
+RUN chown -R node. /opt/learninglocker
```
I would have preferred avoiding this by untaring to the appropriate location using `-C` and by switching to the `node` user before.

jmaupetit commented on nginx/entrypoint.sh:
```diff
@@ -0,0 +1,8 @@
+export DOLLAR='$'
+envsubst < /etc/nginx/conf.d/default.conf.template > /etc/nginx/conf.d/default.conf
+
+exec "$@"
```
If we decide to remove the nginx image, this is no longer required.

jmaupetit commented:
Also I would move the learninglocker/Dockerfile to the project's root and configuration files to `docker/files/`

lunika committed:
📝(Readme) write readme explaining how to build ans use images

jmaupetit commented on Dockerfile:
```diff
@@ -0,0 +1,31 @@
+
+USER node
+
+# Download release and extract it in /opt/learninglocker
```
```suggestion
# Download release and extract it in /app
```

jmaupetit commented on Dockerfile:
```diff
@@ -0,0 +1,31 @@
+
+# Download release and extract it in /opt/learninglocker
+RUN cd /tmp && curl --compressed -L -o "learninglocker-${LL_VERSION}.tar.gz" "https://github.com/LearningLocker/learninglocker/archive/${LL_VERSION}.tar.gz" \
+    && tar -xzf "learninglocker-${LL_VERSION}.tar.gz" -C /app --strip-components 1\
```
```suggestion
    && tar -xzf "learninglocker-${LL_VERSION}.tar.gz" -C /app --strip-components 1 \
```

jmaupetit replied:
:+1:

jmaupetit commented on Dockerfile:
```diff
@@ -0,0 +1,31 @@
+    && rm learninglocker-*.tar.gz
+
+# Install learning locker
+# .env file is created to avoid warning message when learning locker try to parse .env content
```
Please add something like:

> In a Docker context, LearningLocker's environment is injected in the container itself, so theoretically this file is not required. The .env file is only created to avoid application warning message.

jmaupetit commented on README.md:
```diff
@@ -1 +1,45 @@
-# Learning locker docker
+# Learning locker docker images
+
+This repo contains the Dockerfile to build images for [learning locker](https://github.com/LearningLocker/learninglocker), a LRS to store and analyze xAPI statements.
```
```suggestion
This repository contains the `Dockerfile` to build images for [learning locker](https://github.com/LearningLocker/learninglocker), a LRS to store and analyze xAPI statements.
```

jmaupetit commented on README.md:
```diff
@@ -1 +1,45 @@
+
+## Use learning locker image
+
+You have to configure your learning locker image using environment variable. You can found all the variable available in the [documentation](http://docs.learninglocker.net/guides-configuring/#learning-locker-application).
```
```suggestion
You have to configure your learning locker image using environment variables. You will found all available variables in the [project documentation](http://docs.learninglocker.net/guides-configuring/#learning-locker-application).
```

jmaupetit commented on README.md:
```diff
@@ -1 +1,45 @@
+
+You also have to use the [xapi-service](https://github.com/LearningLocker/xapi-service) provided by learning locker. They also provide a docker image you can use.
+
+There is no entrypoint nor command configured in this image. You will have to declare them when you run your container. Here are the command to use:
```
```suggestion
There is no entrypoint nor command configured in this image. You will have to declare them when you run your container. Here are the commands to use:
```

jmaupetit commented on README.md:
````diff
@@ -1 +1,45 @@
+Running the UI application:
+
+```
+node ui/dist/server
````
I usually like to add a shell prompt so that we understand that those are shell commands. WDYT?

```suggestion
$ node ui/dist/server
```

lunika replied:
Yes but you don't execute it in a shell. You will add it in a `command` instruction, this is why I didn't put `$` at the beginning

jmaupetit replied:
True that. In this case, what about the following:

```Dockerfile
# Dockerfile
FROM fundocker/learninglocker:v2.6.2

RUN node ui/dist/server
```

or

```bash
$ docker-run --rm fundocker/learninglocker:v2.6.2 node ui/dist/server
```

jmaupetit commented on README.md:
````diff
@@ -1 +1,45 @@
+Running the API application:
+
+```
+node api/dist/server
````
```suggestion
$ node api/dist/server
```

jmaupetit commented on README.md:
````diff
@@ -1 +1,45 @@
+Running the worker application:
+
+```
+node worker/dist/server
````
```suggestion
$ node worker/dist/server
```

jmaupetit commented on README.md:
````diff
@@ -1 +1,45 @@
+You can also use the CLI by executing the following command from your container:
+
+```
+node cli/dist/server
````
```suggestion
$ node cli/dist/server
```

jmaupetit reviewed (commented):
:clap:

lunika committed:
📝(Readme) add a section explaining docker-compose config

The docker-compose config is here as an example in the repository. We
added a new section in the Readme explaining how to use this example.

jmaupetit commented on example/docker-compose.yml:
```diff
@@ -0,0 +1,77 @@
+
+  xapi:
+    image: learninglocker/xapi-service:2.2.14
+    restart: unless-stopped
```
It seems the only service with a restart policy. Should be removed IMO.

lunika committed:
🔧(circle-ci) configure circle-ci

Everytime a new tag is added we want to publish a new image to docker
hub
